File: pyorion/api/webview.py
# Copyright 2025-2030 Ari Bermeki @ YellowSiC within The Commons Conservancy
# SPDX-License-Identifier: Apache-2.0
# SPDX-License-Identifier: MIT

"""WebView API - Control operations for PyOrion WebView instances.

Wrapper around the Rust event loop backend to manage developer tools
and WebView state.
"""

from pyorion.runtime.runtime_handle import event_register


class WebView:
    """High-level asynchronous wrapper for controlling a WebView instance."""

    def __init__(self) -> None:
        """Initialize a new :class:`WebView` handle."""

    async def is_devtools_open(self) -> bool:
        """Check if the WebView developer tools window is currently open.

        :return: ``True`` if DevTools is open, otherwise ``False``.
        :rtype: bool
        """
        return await event_register("webview.isDevtoolsOpen", None, result_type=bool)

    async def open_devtools(self) -> bool:
        """Open the WebView developer tools window.

        :return: ``True`` if the operation succeeded, otherwise ``False``.
        :rtype: bool
        """
        return await event_register("webview.openDevtools", None, result_type=bool)

    async def close_devtools(self) -> bool:
        """Close the WebView developer tools window.

        :return: ``True`` if the operation succeeded, otherwise ``False``.
        :rtype: bool
        """
        return await event_register("webview.closeDevtools", None, result_type=bool)

File: src/api/webview.rs
// Copyright 2025-2030 Ari Bermeki @ YellowSiC within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

use anyhow::Result;
use pyorion_macros::api;

use crate::api_manager::ApiManager;

pub fn webview_api(api_manager: &mut ApiManager) {
    api_manager.register_api("webview.isDevtoolsOpen", is_devtools_open);
    api_manager.register_api("webview.openDevtools", open_devtools);
    api_manager.register_api("webview.closeDevtools", close_devtools);
}

#[api]
fn is_devtools_open() -> Result<bool> {
    let webview = app.app_context()?.get_webview()?;
    Ok(webview.is_devtools_open())
}

#[api]
fn open_devtools() -> Result<()> {
    let webview = app.app_context()?.get_webview()?;
    webview.open_devtools();
    Ok(())
}

#[api]
fn close_devtools() -> Result<()> {
    let webview = app.app_context()?.get_webview()?;
    webview.close_devtools();
    Ok(())
}
